### ISSUE TITLE
False positive [E602] Don't compare to empty string

### ISSUE THREAD

**Issue by ssbarnea** [ansible/ansible-lint#450]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

```
ansible --version
2.5.7
ansible-lint --version
4.0.0
```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

No false positive errors

# Actual Behaviour (Bug report only)

    [E602] Don't compare to empty string

See https://github.com/openstack/tripleo-quickstart/blob/master/roles/tripleo-inventory/tasks/openstack.yml#L60

I found 4 similar places where bit more complex jinja2 syntax triggers this bug which is weird as there is no conditional involved here.

Any idea on how to avoid this until it is fixed?

**Comment by geerlingguy** [comment]
So I just came here to note that I am not sure what to do instead of comparing to an empty string, since the language doesn't support strict types and this is the only way to check if a string is 'empty' or 'not empty'.

For example, with Ansible 2.7.x, if I set `my_var: 'testing'` and have the following task:

```
- name: Do something with my_var.
  command: "do_something_with {{ my_var }}"
  when: my_var
```

Then I get the error:

```
fatal: [localhost]: FAILED! =>
The conditional check 'my_var' failed. The error was: error while evaluating conditional (my_var): 'testing' is undefined
```

If I set it like:

```
- name: Do something with my_var.
  command: "do_something_with {{ my_var }}"
  when: my_var != ''
```

Then I have no problem, and it is more correct IMO.

So basically, if that's wrong... what's the right way to do that kind of check? I am scratching my head and I use this pattern fairly often.

**Comment by sseide** [comment]
Same problem for me - using conditional check as ansible-lint (and ansible-galaxy) wants the playbook exits with an error. 
Using the example from Jeff Geerling above - my_var is set to `'/var/run/myapp.pid'` . When task is run with conditional `when: my_var` ansible exits with error:

```
fatal: [instance]: FAILED! => {"msg": "The conditional check 'my_var' failed. 
The error was: template error while templating string: unexpected '/'. 
String: {% if /var/run/myapp.pid %} True {% else %} False {% endif %}\n\nThe error appears
to have been in '/bla/blub.yml': line 14, column 3, but may\nbe elsewhere in the file 
depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- 
name: \"Do something with my_var\"\n  ^ here\n"}
```

How do we check if `when: my_var != ''` is not allowed?

**Comment by awcrosby** [comment]
The false positive was matching on `=='"` and the referenced PR will fix this by updating the regex.

As far as the other comments, for those that need to use this pattern, hopefully the addition of granular skipping planned in the next feature release will address most concerns - and those that find the rule doesn't work for them, can skip this rule entirely via [config file](https://docs.ansible.com/ansible-lint/configuring/configuring.html#id2) or [command line option](https://docs.ansible.com/ansible-lint/usage/usage.html#id2)

**Comment by geerlingguy** [comment]
@awcrosby - Should @sseide and myself open a new ticket, then? The problem is, there are _tons_ of valid use cases for comparing to empty strings... especially in a duck-typed language like Python, and I have no idea how to cleanly do things any differently.

I hit this rule on probably 30% of my roles and playbooks, and for now I'm just disabling the rule whenever I encounter it.

I'd like to know what the rationale is behind _not_ comparing to empty strings, ever.

**Comment by geerlingguy** [comment]
@awcrosby yes, that's what I'm currently doing, but I feel terrible doing that because I'm a bit of a purist and would like to agree with lint rules and not ignore them :D

**Comment by geerlingguy** [comment]
I guess another option which would at least work around this would be to use the Jinja2 `length` filter instead... it's doing the same thing but IMO is more obtuse, but at least it would pass the linter without throwing an error.

    when: mysql_replication_master != ''
    when: mysql_replication_master | length > 0

It just seems like an anti-pattern to do the same thing but in a way that makes less semantic sense because of a lint rule. I favor code readability in this case.

**Comment by geerlingguy** [comment]
I opened https://github.com/ansible/ansible-lint/issues/457 to discuss the rule further.

**Comment by sseide** [comment]
As Jeff Geerling found out (thanks!) the problem is related to a old bug report in ansible itself (ansible/ansible#39414)
Whenever this gets fixed this one will be solved too...

**Comment by ericsysmin** [comment]
The lint here is annoying for sure @geerlingguy haha

**Comment by shanhanqiang** [comment]
The funny thing here is that in the official documents it's exactly comparing to an empty string.
https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html#conditions-based-on-registered-variables
